Nested Show Errors:

As soon as I reference a show from within another show like this (myparentshow.yaml): 

```
#show_version=4
- time: 0
  shows:
    mychildshow:
      speed: 1
      show_tokens:
        line1num: row0
```

I am getting the following error:

```
MPF Shutting down due to child thread crash
Crash details: %s Traceback (most recent call last):
  File "/Applications/MPF.app/Contents/Frameworks/python/3.5.0/lib/python3.5/site-packages/mpf/core/assets.py", line 622, in run
    asset.do_load()
  File "/Applications/MPF.app/Contents/Frameworks/python/3.5.0/lib/python3.5/site-packages/mpf/assets/show.py", line 83, in do_load
    self._do_load_show(None)
  File "/Applications/MPF.app/Contents/Frameworks/python/3.5.0/lib/python3.5/site-packages/mpf/assets/show.py", line 128, in _do_load_show
    self._show_validation_error("Cannot load empty show")
  File "/Applications/MPF.app/Contents/Frameworks/python/3.5.0/lib/python3.5/site-packages/mpf/assets/show.py", line 187, in _show_validation_error
    raise AssertionError("Show {}: {}".format(identifier, msg))
AssertionError: Show /Users/myusername/Documents/pinball/training/shows/myparentshow.yaml: Cannot load empty show

INFO : Machine : Shutting down...
INFO : Machine : Actual MPF loop rate: 0 Hz
ERROR : root : 'NoneType' object has no attribute 'close'
Traceback (most recent call last):
  File "/Applications/MPF.app/Contents/Frameworks/python/3.5.0/lib/python3.5/site-packages/mpf/commands/game.py", line 130, in __init__
    MachineController(mpf_path, machine_path, vars(args)).run()
  File "/Applications/MPF.app/Contents/Frameworks/python/3.5.0/lib/python3.5/site-packages/mpf/core/machine.py", line 540, in run
    self._run_loop()
  File "/Applications/MPF.app/Contents/Frameworks/python/3.5.0/lib/python3.5/site-packages/mpf/core/machine.py", line 568, in _run_loop
    self._platform_stop()
  File "/Applications/MPF.app/Contents/Frameworks/python/3.5.0/lib/python3.5/site-packages/mpf/core/machine.py", line 588, in _platform_stop
    platform.stop()
  File "/Applications/MPF.app/Contents/Frameworks/python/3.5.0/lib/python3.5/site-packages/mpf/platforms/fast/fast.py", line 123, in stop
    connection.stop()
  File "/Applications/MPF.app/Contents/Frameworks/python/3.5.0/lib/python3.5/site-packages/mpf/platforms/fast/fast.py", line 815, in stop
    self.serial_connection.close()
AttributeError: 'NoneType' object has no attribute 'close'
```

Additionally, if myparentshow references any other shows, I get the same error even if myparentshow is not in use. I cannot start the machine successfully until myparentshow.yaml is moved out of the shows folder to somewhere innocuous.

--- a/mpf/assets/show.py
+++ b/mpf/assets/show.py
@@ -376,7 +376,6 @@ class Show(Asset):
                                    loops=int(loops),
                                    sync_ms=int(sync_ms),
                                    reset=bool(reset),
-                                   mode=mode,
                                    manual_advance=manual_advance,
                                    show_tokens=show_tokens)
 
@@ -408,7 +407,7 @@ class RunningShow(object):
     # pylint: disable-msg=too-many-locals
     def __init__(self, machine, show, show_steps, priority,
                  speed, start_step, callback, loops,
-                 sync_ms, reset, mode, manual_advance, show_tokens):
+                 sync_ms, reset, manual_advance, show_tokens):
         """Initialise an instance of a show."""
         self.machine = machine
         self.show = show
@@ -427,8 +426,6 @@ class RunningShow(object):
 
         self.next_step_time = self.machine.clock.get_time()
 
-        del mode
-        # TODO: remove mode from __init__
         self.manual_advance = manual_advance
 
         self.name = show.name

--- a/mpf/assets/show.py
+++ b/mpf/assets/show.py
@@ -339,6 +339,7 @@ class Show(Asset):
             the show is not loaded. (In this case the show will be loaded and
             will automatically play once its loaded.)
         """
+        del mode
         # todo bugfix, currently there is only one set of autoplay seetings,
         # so if multiple show instances are played but the show is not loaded,
         # only the last one will play

--- a/mpf/assets/show.py
+++ b/mpf/assets/show.py
@@ -423,6 +423,9 @@ class RunningShow(object):
         # self.mode = mode
         self.show_tokens = show_tokens
         self._delay_handler = None
+        self.next_step_index = None
+
+        self.next_step_time = self.machine.clock.get_time()
 
         del mode
         # TODO: remove mode from __init__
@@ -448,6 +451,11 @@ class RunningShow(object):
             self._show_loaded = False
 
     def show_loaded(self, show):
+        """Called when a deferred show was loaded.
+
+        Start playing the show as if it started earlier.
+        """
+        del show
         self._show_loaded = True
         self.show_steps = self.show.get_show_steps()
         self._start_play()
@@ -472,8 +480,6 @@ class RunningShow(object):
         self.machine.show_controller.notify_show_starting(self)
 
         # Figure out the show start time
-        self.next_step_time = self.machine.clock.get_time()
-
         if self.sync_ms:
             delay_secs = (self.sync_ms / 1000.0) - (self.next_step_time % (self.sync_ms / 1000.0))
             self.next_step_time += delay_secs

--- a/mpf/assets/show.py
+++ b/mpf/assets/show.py
@@ -361,40 +361,29 @@ class Show(Asset):
                              'expected: {}. Tokens submitted: {}'.
                              format(self.name, self.tokens, set(show_tokens.keys())))
 
+        if self.loaded:
+            show_steps = self.get_show_steps()
+        else:
+            show_steps = False
+
+        running_show = RunningShow(machine=self.machine,
+                                   show=self,
+                                   show_steps=show_steps,
+                                   priority=int(priority),
+                                   speed=float(speed),
+                                   start_step=int(start_step),
+                                   callback=callback,
+                                   loops=int(loops),
+                                   sync_ms=int(sync_ms),
+                                   reset=bool(reset),
+                                   mode=mode,
+                                   manual_advance=manual_advance,
+                                   show_tokens=show_tokens)
+
         if not self.loaded:
-            self._autoplay_settings = dict(priority=priority,
-                                           speed=speed,
-                                           start_step=start_step,
-                                           callback=callback,
-                                           loops=loops,
-                                           sync_ms=sync_ms,
-                                           reset=reset,
-                                           mode=mode,
-                                           manual_advance=manual_advance,
-                                           show_tokens=show_tokens
-                                           )
+            self.load(callback=running_show.show_loaded, priority=priority)
 
-            self.load(callback=self._autoplay, priority=priority)
-            return False
-
-        return RunningShow(machine=self.machine,
-                           show=self,
-                           show_steps=self.get_show_steps(),
-                           priority=int(priority),
-                           speed=float(speed),
-                           start_step=int(start_step),
-                           callback=callback,
-                           loops=int(loops),
-                           sync_ms=int(sync_ms),
-                           reset=bool(reset),
-                           mode=mode,
-                           manual_advance=manual_advance,
-                           show_tokens=show_tokens)
-
-    def _autoplay(self, *args, **kwargs):
-        del args
-        del kwargs
-        self.play(**self._autoplay_settings)
+        return running_show
 
     def load_show_from_disk(self):
         """Load show from disk."""
@@ -429,6 +418,8 @@ class RunningShow(object):
         self.callback = callback
         self.loops = loops
         self.reset = reset
+        self.start_step = start_step
+        self.sync_ms = sync_ms
         # self.mode = mode
         self.show_tokens = show_tokens
         self._delay_handler = None
@@ -450,27 +441,41 @@ class RunningShow(object):
         self.debug = False
         self._stopped = False
 
-        self._total_steps = len(show_steps)
+        if show_steps:
+            self._show_loaded = True
+            self._start_play()
+        else:
+            self._show_loaded = False
 
-        if start_step > 0:
-            self.next_step_index = start_step - 1
-        elif start_step < 0:
-            self.next_step_index = self._total_steps + start_step
+    def show_loaded(self, show):
+        self._show_loaded = True
+        self.show_steps = self.show.get_show_steps()
+        self._start_play()
+
+    def _start_play(self):
+        if self._stopped:
+            return
+
+        self._total_steps = len(self.show_steps)
+
+        if self.start_step > 0:
+            self.next_step_index = self.start_step - 1
+        elif self.start_step < 0:
+            self.next_step_index = self._total_steps + self.start_step
         else:
             self.next_step_index = 0
 
-        if show_tokens and show.tokens:
-            self._replace_tokens(**show_tokens)
+        if self.show_tokens and self.show.tokens:
+            self._replace_tokens(**self.show_tokens)
 
-        show.running.add(self)
+        self.show.running.add(self)
         self.machine.show_controller.notify_show_starting(self)
 
         # Figure out the show start time
         self.next_step_time = self.machine.clock.get_time()
 
-        if sync_ms:
-            delay_secs = (sync_ms / 1000.0) - (self.next_step_time % (sync_ms /
-                                               1000.0))
+        if self.sync_ms:
+            delay_secs = (self.sync_ms / 1000.0) - (self.next_step_time % (self.sync_ms / 1000.0))
             self.next_step_time += delay_secs
             self._delay_handler = self.machine.clock.schedule_once(self._run_next_step,
                                                                    delay_secs)
@@ -524,6 +529,10 @@ class RunningShow(object):
             return
 
         self._stopped = True
+
+        if not self._show_loaded:
+            return
+
         # todo I think there's a potential memory leak here as shows stop but
         # some things still hold references to them. (Shots, for example).
         # need to investigate more
@@ -549,6 +558,8 @@ class RunningShow(object):
 
     def resume(self):
         """Resume paused show."""
+        if not self._show_loaded:
+            return
         self.next_step_time = self.machine.clock.get_time()
         self._run_next_step()
 
@@ -563,6 +574,9 @@ class RunningShow(object):
 
     def advance(self, steps=1, show_step=None):
         """Manually advance this show to the next step."""
+        if not self._show_loaded:
+            # todo: this is suboptimal
+            return
         if isinstance(show_step, int) and show_step < 0:    # pragma: no cover
             raise ValueError('Cannot advance {} to step "{}" as that is'
                              'not a valid step number.'.format(self, show_step))

--- a/mpf/core/assets.py
+++ b/mpf/core/assets.py
@@ -16,15 +16,15 @@ from mpf.core.utility_functions import Util
 
 
 class AssetManager(object):
+
     """Base class for the Asset Manager.
 
     Args:
-        mpfmc: The main MpfMc object.
-
+        machine: The machine controller
     """
 
     def __init__(self, machine):
-
+        """Initialise asset manager."""
         self.log = logging.getLogger('AssetManager')
         self.log.debug("Initializing...")
 
@@ -67,8 +67,9 @@ class AssetManager(object):
 
     @property
     def loading_percent(self):
-        """The percent of assets that are in the process of loading that have
-        been loaded. This value is an integer between 0 and 100. It's reset
+        """Return the percent of assets that are in the process of loading that have been loaded.
+
+        This value is an integer between 0 and 100. It's reset
         when all the assets have been loaded, so it will go from 0 to 100 when
         MPF is starting up, and then go from 0 to 100 again when a mode starts,
         etc.
@@ -76,7 +77,6 @@ class AssetManager(object):
         Note that this percentage also includes asset loading status updates
         from a connected BCP client.
         """
-
         try:
             return round((self.num_assets_loaded +
                           self.num_bcp_assets_loaded) /
@@ -85,15 +85,6 @@ class AssetManager(object):
 
         except ZeroDivisionError:
             return 100
-
-    # def shutdown(self):
-    #     """Prepares the Asset Manager for shutdown by stopping the loader
-    #     thread. Will block until the loader thread is stopped.
-    #
-    #     """
-    #     self.loader_queue = None
-    #     self.loaded_queue = None
-    #     self.loader_thread.stop()
 
     def _start_loader_thread(self):
         self.loader_thread = AssetLoader(loader_queue=self.loader_queue,
@@ -108,7 +99,7 @@ class AssetManager(object):
                              disk_asset_section,
                              path_string, extensions, priority,
                              pool_config_section):
-        """Registers a a type of assets to be controlled by the AssetManager.
+        """Register a a type of assets to be controlled by the AssetManager.
 
         Args:
             asset_class: Reference to the class you want to register, based on
@@ -211,8 +202,9 @@ class AssetManager(object):
             self.machine.clear_boot_hold('assets')
 
     def _create_assets_from_disk(self, config, mode=None):
-        """Walks a folder (and subfolders) and finds all the assets. Checks to
-        see if those assets have config entries in the passed config file, and
+        """Walk a folder (and subfolders) and finds all the assets.
+
+        Check to see if those assets have config entries in the passed config file, and
         then builds a config for each asset based on its config entry, and/or
         defaults based on the subfolder it was in or the general defaults.
         Then it creates the asset objects based on the built-up config.
@@ -272,7 +264,6 @@ class AssetManager(object):
         every asset it found in that folder and subfolders (along with its
         full path), and a config dict appropriately merged from default,
         folder-specific, and asset specific settings
-
         """
         if not config:
             config = dict()
@@ -309,8 +300,9 @@ class AssetManager(object):
     # pylint: disable-msg=too-many-locals
     def _create_asset_config_entries(self, asset_class, config, mode_name=None,
                                      path=None):
-        """Scans a folder (and subfolders) and automatically creates or updates
-        entries in the config dict for any asset files it finds.
+        """Scan a folder (and subfolders).
+
+        Automatically creates or updates entries in the config dict for any asset files it finds.
 
         Args:
             asset_class: An asset class entry from the self._asset_classes
@@ -354,7 +346,6 @@ class AssetManager(object):
                 /images/foo/image5.png
                 /images/other/images6.png
                 /images/other/big/image7.png
-
         """
         if not path:
             path = self.machine.machine_path
@@ -440,7 +431,7 @@ class AssetManager(object):
                     priority=priority))
 
     def load_assets_by_load_key(self, key_name, priority=0):
-        """Loads all the assets with a given load key.
+        """Load all the assets with a given load key.
 
             Args:
                 key_name: String of the load: key name.
@@ -461,7 +452,7 @@ class AssetManager(object):
 
     @classmethod
     def unload_assets(cls, assets):
-        """Unloads multiple assets.
+        """Unload multiple assets.
 
             Args:
                 assets: An iterable of asset objects. You can safely mix
@@ -472,6 +463,7 @@ class AssetManager(object):
             asset.unload()
 
     def load_asset(self, asset):
+        """Put asset in loader queue."""
         # Internal method which handles the logistics of actually loading an
         # asset. Should only be called by Asset.load() as that method does
         # additional things that are needed.
@@ -524,7 +516,6 @@ class AssetManager(object):
             loaded=self.num_assets_loaded + self.num_bcp_assets_loaded,
             remaining=remaining,
             percent=self.loading_percent)
-
         '''event: loading_assets
 
         desc: Posted when the number of assets waiting to be loaded changes.
@@ -578,8 +569,8 @@ class AssetManager(object):
 
 
 class AssetLoader(threading.Thread):
-    """Base class for the Asset Loader thread and actually loads the assets
-    from disk.
+
+    """Base class for the Asset Loader thread and actually loads the assets from disk.
 
     Args:
         loader_queue: A reference to the asset manager's loader_queue which
@@ -592,12 +583,11 @@ class AssetLoader(threading.Thread):
             the asset loader crashes, it will write the crash to that queue and
             cause an exception in the main thread. Otherwise it fails silently
             which is super annoying. :)
-
     """
 
     def __init__(self, loader_queue, loaded_queue, exception_queue,
                  thread_stopper):
-
+        """Initialise asset loader."""
         threading.Thread.__init__(self)
         self.log = logging.getLogger('Asset Loader')
         self.loader_queue = loader_queue
@@ -633,7 +623,11 @@ class AssetLoader(threading.Thread):
 
 
 class AssetPool(object):
+
+    """Pool of assets."""
+
     def __init__(self, mc, name, config, member_cls):
+        """Initialise asset pool."""
         self.machine = mc
         self.priority = None
         self.name = name
@@ -678,6 +672,7 @@ class AssetPool(object):
 
     @property
     def asset(self):
+        """Pop one asset from the pool."""
         if self.config['type'] == 'random':
             return self._get_random_asset()
         elif self.config['type'] == 'sequence':
@@ -697,6 +692,7 @@ class AssetPool(object):
             self._asset_sequence.rotate(1)
 
     def load(self, callback=None, priority=None):
+        """Load pool."""
         if priority is not None:
             self.priority = priority
 
@@ -763,6 +759,9 @@ class AssetPool(object):
 
 
 class Asset(object):
+
+    """Baseclass for all assets."""
+
     attribute = ''  # attribute in MC, e.g. self.machine.images
     path_string = ''  # entry from mpf-mc:paths: for asset folder name
     config_section = ''  # section in the config files for this asset
@@ -785,12 +784,12 @@ class Asset(object):
         return cls._next_id
 
     @classmethod
-    def initialize(cls, mc):
-
+    def initialize(cls, machine):
+        """Initialise asset class."""
         if not cls.disk_asset_section:
             cls.disk_asset_section = cls.config_section
 
-        mc.asset_manager.register_asset_class(
+        machine.asset_manager.register_asset_class(
             asset_class=cls,
             attribute=cls.attribute,
             path_string=cls.path_string,
@@ -800,8 +799,9 @@ class Asset(object):
             priority=cls.class_priority,
             pool_config_section=cls.pool_config_section)
 
-    def __init__(self, mc, name, file, config):
-        self.machine = mc
+    def __init__(self, machine, name, file, config):
+        """Initialise asset."""
+        self.machine = machine
         self.name = name
         self.file = file
 
@@ -823,15 +823,18 @@ class Asset(object):
                                                   self.name, self.loaded)
 
     def __lt__(self, other):
+        """Compare assets."""
         # Note this is "backwards" (It's the __lt__ method but the formula uses
         # greater than because the PriorityQueue puts lowest first.)
         return ("%s, %s" % (self.priority, self._id) >
                 "%s, %s" % (other.priority, other.get_id()))
 
     def get_id(self):
+        """Return id."""
         return self._id
 
     def load(self, callback=None, priority=None):
+        """Start loading the asset."""
         if priority is not None:
             self.priority = priority
 
@@ -856,18 +859,21 @@ class Asset(object):
         self._callbacks = set()
 
     def do_load(self):
+        """Load the asset blocking."""
         # This is the actual method that loads the asset. It's called by a
         # different thread so it's ok to block. Make sure you don't set any
         # attributes here or you don't need any since it's a separate thread.
         raise NotImplementedError
 
     def is_loaded(self):
+        """Called when asset has been loaded."""
         self.loading = False
         self.loaded = True
         self.unloading = False
         self._call_callbacks()
 
     def unload(self):
+        """Called when the asset has been unloaded."""
         self.unloading = True
         self.loaded = False
         self.loading = False

--- a/mpf/core/assets.py
+++ b/mpf/core/assets.py
@@ -54,6 +54,9 @@ class AssetManager(object):
 
         self._start_loader_thread()
 
+        self._next_id = 0
+        # id of next asset
+
         self.machine.mode_controller.register_start_method(
             start_method=self._load_mode_assets)
 
@@ -64,6 +67,11 @@ class AssetManager(object):
         # Picks up asset load information from connected BCP client(s)
         self.machine.events.add_handler('assets_to_load',
                                         self._bcp_client_asset_load)
+
+    def get_next_id(self):
+        """Return the next free id."""
+        self._next_id += 1
+        return self._next_id
 
     @property
     def loading_percent(self):
@@ -771,18 +779,6 @@ class Asset(object):
     pool_config_section = None  # Create an associated AssetPool instance
     asset_group_class = AssetPool  # replace with your own asset group class
 
-    _next_id = 0
-
-    @classmethod
-    def _get_id(cls):
-        # Since the asset loader priority queue needs a way to break ties if
-        # two assets are loading with the same priority, we need to implement
-        # a comparison operator on the Asset, so we just increment and ID.
-        # This means the assets will load in the order they were added to the
-        # queue
-        cls._next_id += 1
-        return cls._next_id
-
     @classmethod
     def initialize(cls, machine):
         """Initialise asset class."""
@@ -811,7 +807,7 @@ class Asset(object):
 
         self.priority = self.config.get('priority', 0)
         self._callbacks = set()
-        self._id = Asset._get_id()
+        self._id = machine.asset_manager.get_next_id()
         self.lock = threading.Lock()
 
         self.loading = False  # Is this asset in the process of loading?

--- a/mpf/core/case_insensitive_dict.py
+++ b/mpf/core/case_insensitive_dict.py
@@ -1,37 +1,51 @@
+"""Case insensitive dict."""
 # Based on this: http://stackoverflow.com/questions/2082152/case-insensitive-dictionary
 
 
 class CaseInsensitiveDict(dict):
+
+    """A dict which lowercases all keys."""
+
     @staticmethod
     def lower(key):
+        """Lowercase the key."""
         return key.lower() if isinstance(key, str) else key
 
     def __init__(self, *args, **kwargs):
+        """Initialise case insensitve dict."""
         super().__init__(*args, **kwargs)
         self._convert_keys()
 
     def __getitem__(self, key):
+        """Return item for key."""
         return super().__getitem__(self.__class__.lower(key))
 
     def __setitem__(self, key, value):
+        """Set item for key to value."""
         super().__setitem__(self.__class__.lower(key), value)
 
     def __delitem__(self, key):
+        """Delete item for key."""
         return super().__delitem__(self.__class__.lower(key))
 
     def __contains__(self, key):
+        """Check if dict contains a key."""
         return super().__contains__(self.__class__.lower(key))
 
     def pop(self, key, *args, **kwargs):
+        """Retrieve and delete a value for a key."""
         return super().pop(self.__class__.lower(key), *args, **kwargs)
 
     def get(self, key, *args, **kwargs):
+        """Return item for key."""
         return super().get(self.__class__.lower(key), *args, **kwargs)
 
     def setdefault(self, key, *args, **kwargs):
+        """Set defaults."""
         return super().setdefault(self.__class__.lower(key), *args, **kwargs)
 
     def update(self, e=None, **f):
+        """Update a value for a key."""
         super().update(self.__class__(e))
         super().update(self.__class__(**f))
 

--- a/mpf/tests/MpfTestCase.py
+++ b/mpf/tests/MpfTestCase.py
@@ -138,6 +138,7 @@ class MpfTestCase(unittest.TestCase):
 
     def advance_time_and_run(self, delta=1.0):
         self.machine.log.info("Advancing time by %s", delta)
+        e = False
         try:
             self.loop.run_until_complete(asyncio.sleep(delay=delta, loop=self.loop))
             return
@@ -146,8 +147,10 @@ class MpfTestCase(unittest.TestCase):
 
         if self._exception:
             raise self._exception['exception']
-        else:
+        elif e:
             raise e
+        else:
+            raise AssertionError("error")
 
     def machine_run(self):
         self.advance_time_and_run(0)

--- a/mpf/tests/machine_files/shows/config/test_shows.yaml
+++ b/mpf/tests/machine_files/shows/config/test_shows.yaml
@@ -49,15 +49,6 @@ flashers:
         label: Test flasher
         flash_ms: 40
 
-assets:
-    file_shows:
-        test_show1:
-            preload: yes
-        test_show2:
-            preload: yes
-        test_show3:
-            preload: yes
-
 shows:
   leds_name_token:
     - time: 0

--- a/mpf/tests/machine_files/shows/shows/myparentshow.yaml
+++ b/mpf/tests/machine_files/shows/shows/myparentshow.yaml
@@ -1,0 +1,6 @@
+#show_version=4
+- duration: -1
+  shows:
+    mychildshow:
+      speed: 1
+      loops: 0

--- a/mpf/tests/machine_files/shows/shows/on_demand/mychildshow.yaml
+++ b/mpf/tests/machine_files/shows/shows/on_demand/mychildshow.yaml
@@ -1,0 +1,3 @@
+#show_version=4
+- time: 0
+  events: test

--- a/mpf/tests/test_Shows.py
+++ b/mpf/tests/test_Shows.py
@@ -703,3 +703,27 @@ class TestShows(MpfTestCase):
         self.machine.shows['multiple_tokens'].play(show_tokens=dict(
             lights='light_01'))
 
+    def test_nested_shows(self):
+        self.mock_event("test")
+        self.assertFalse(self.machine.shows['mychildshow'].loaded)
+        show = self.machine.shows['myparentshow'].play(loops=0)
+
+        self.advance_time_and_run(5)
+        while not self.machine.shows['mychildshow'].loaded:
+            self.advance_time_and_run(1)
+        self.assertEqual(1, self._events['test'])
+        self.assertTrue(self.machine.shows['mychildshow'].loaded)
+        show.stop()
+
+    def test_nested_shows_stop_before_load(self):
+        self.mock_event("test")
+        self.assertFalse(self.machine.shows['mychildshow'].loaded)
+        show = self.machine.shows['myparentshow'].play(loops=0)
+        show.stop()
+        self.assertEqual(0, self._events['test'])
+
+        self.advance_time_and_run(5)
+        while not self.machine.shows['mychildshow'].loaded:
+            self.advance_time_and_run(1)
+        self.assertEqual(0, self._events['test'])
+        self.assertTrue(self.machine.shows['mychildshow'].loaded)

